### PR TITLE
TYP: fix ignores

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/dtypes/cast.py
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:556]: could perhaps create a _DtypeObjT typvar (local to the module if we don't need elsewhere) instead of an overload here since ensure_dtype_can_hold_na is type-preserving

File: pandas/core/arrays/datetimes.py
Comment (simonjayhawkins) [inline pandas/core/arrays/datetimes.py:1928]: with these Literal[True]/Literal[False] may also need to add an extra bool case. ie. duplicate the function signature as an additional overload (follow-on ok as may not be needed if this is internal) see #40200 and #40197

File: pandas/core/dtypes/cast.py
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:410]: The logic has changed here slightly. is there a latent bug here that needs a test.

File: pandas/core/dtypes/cast.py
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:556]: i considered that and ended up not pulling the trigger, would be OK with it if you feel strongly about it

File: pandas/core/arrays/datetimes.py
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py:1928]: > may not be needed if this is internal

yah this shouldnt be user-facing

File: pandas/core/dtypes/cast.py
Comment (simonjayhawkins) [reply pandas/core/dtypes/cast.py:410]: looking some more, maybe the type annotations for maybe_downcast_to_dtype are incorrect?

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:410]: since this is only called from one place and that is with numeric_only=True (ought to remove that kwarg) and the two cases excluded here dont have is_numeric_dtype(dtype), the behavior will end up being the same before and after

Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:410]: > looking some more, maybe the type annotations for maybe_downcast_to_dtype are incorrect?

it does look sketchy, yah

File: pandas/core/dtypes/cast.py
Comment (simonjayhawkins) [reply pandas/core/dtypes/cast.py:556]: > if you feel strongly about it

I wouldn't block on this, so i'll leave it to you.

I won't be able to fully review this PR till tomorrow, so the comments so far are just some bits that stood out.

**Review by simonjayhawkins — APPROVED** [review]
lgtm. making good progress clearing down the numpy related ignores. Thanks @jbrockmendel .

File: pandas/core/dtypes/cast.py
Comment (jbrockmendel) [reply pandas/core/dtypes/cast.py:1466]: updated

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**
